target/linux/ramips/mt7621/base-files/lib/upgrade/dna.sh

```diff
@@ -0,0 +1,43 @@
+#
+# Copyright (C) 2023 Mauri Sandberg
+#
+
+# The vendor ubi is split in ubifs volumes 0-3 and the vendor root
+# filesystems are in volumes 2 (rootfs_0) and 3 (rootfs_1). Drop them and
+# explicitly use rootfs_0 as a boot partition that contains the dtb and the
+# OpenWrt kernel. This is because the u-boot expects to find them there.
+# Then continue upgrade with the default method and a squashfs rootfs will be
+# installed and the rest of ubifs will be used as overlay
+
+# The 'kernel' inside the sysupgrage.tar is an ubifs that contains /boot/dtb
+# and /boot/kernel. The 'root' is an OpenWrt squashfs root
+
+. /lib/functions.sh
+. /lib/upgrade/nand.sh
+
+dna_do_upgrade () {
+	tar -xaf $1
+
+	# get the size of the new bootfs
+	local _bootfs_size=$(wc -c < ./sysupgrade-dna_valokuitu-plus-ex400/kernel)
+	[ -n "$_bootfs_size" -a "$_bootfs_size" -gt "0" ] || nand_do_upgrade_failed
+
+	# remove existing rootfses and recreate rootfs_0
+	ubirmvol /dev/ubi0 --name=rootfs_0 > /dev/null 2>&1
+	ubirmvol /dev/ubi0 --name=rootfs_1 > /dev/null 2>&1
+	ubirmvol /dev/ubi0 --name=rootfs > /dev/null 2>&1
+	ubirmvol /dev/ubi0 --name=rootfs_data > /dev/null 2>&1
+	ubimkvol /dev/ubi0 --type=static --size=${_bootfs_size} --name=rootfs_0
```

Comment:
Any reason in particular why you pick a static volume here?

Reply:
Yes, I wanted to make the "boot partition" a read-only.

Reply:
The boot volume is already mounted read-only in OpenWrt. dynamic vs static has other implications. Most notably, static volumes are slower to access (but have less overhead in terms of flash space). It is better to use static volumes only for things that are read and written to in one go like kernel, U-Boot, calibration data. See http://www.linux-mtd.infradead.org/faq/ubi.html#L_dyn_faster for some more detail.

OpenWrt also uses dynamic volumes for rootfs (you can see by looking at the invocations of ubimkvol for other devices and (a bit harder to see) from the invocations of  scripts/ubinize-image.sh via the defines in the Makefiles which until recently did not even support static volumes.

Reply:
The volume `rootfs_0` contains only the kernel and the dtb and I call it loosely `bootfs` as the u-boot expects to find it there and to contain said elements. It's not OpenWrt as such but a method of booting it up and I cannot see how OpenWrt would make it read-only. The read-only squashfs root is in volume `rootfs` and the rest of ubi becomes a dynamic overlay volume called `rootfs_data`.

To my understanding dynamic volumes are writable but static ones are not. I just wanted to make the volume, which is part of booting up, unmutable. On the otherhand, how would you make a dynamic volume read-only?